batcher.go

```diff
@@ -1,5 +1,7 @@
 package batcher
 
+// NOTE: please review this code which organizes operations into batches based on criteria
+
```

Comment:
**Final Notes:**
- For some of the batcher tests, it might be worth leaving comments that call out... 
  - when a `rate limiter` is NOT being used. Comments that point out when batching is not enabled might be helpful as well. (ex. Line 87 has a test which does does not use both rate limiting and batching. Calling out subtle context switches like that would speed up understanding.)
  - when recursion is being used. This should speed up understanding. (ex. Line 87 has a test where `enqueue()` calls `batcher.Enqueue()` just to call `enqueue()` again. This is the test where max-attempts governor is being tested. )
- I see that the [Docs PR](https://github.com/plasne/go-batcher/pull/4) will have a dedicated section for understanding the `FlushInterval`. That is definitely something I was going to ask for. `CapacityInterval` might deserve it's own section as well. I think these are the primary channels pushing the payload to the datastore. IMO, everything else is just managing that data flow.
- Tests coverage would be even closer to 100% if `NewProvisionResource` actually had logic to test. You're probably already aware.
- Nothing is happening with the `count` variable on line 308.

Reply:
- Tests should only do the minimal necessary to test the condition. This means that most tests do not have rate limiting or batching. I think it would be potentially more confusing to call out every feature that is used or not used in a test. Rather, if someone wants to understand a test, reading the code should provide that.
  - I added a note about the recursion in case that was confusing.

- I had actually am removing the FlushInterval guidance because there wasn't any guidance beyond what was already discussed. I really didn't have anything else to say about them other than how they are already described:
  - __FlushInterval__ [DEFAULT: 100ms]: This determines how often Operations in the buffer are examined. Each time the interval fires, Operations will be dequeued and added to batches or released individually (if not batchable) until such time as the aggregate cost of everything considered in the interval exceeds the capacity allotted this timeslice. For the 100ms default, there will be 10 intervals per second, so the capacity allocated is 1/10th the available capacity. Generally you want FlushInterval to be under 1 second though it could technically go higher.
  - __CapacityInterval__ [DEFAULT: 100ms]: This determines how often the Batcher asks the RateLimiter for capacity. Generally you should leave this alone, but you could increase it to slow down the number of storage Operations required for sharing capacity. Please be aware that this only applies to Batcher asking for capacity, it doesn't mean the rate limiter will allocate capacity any faster, just that it is being asked more often.

I added this if it helps...

## Simple Explanation

1. A Batcher is created for a datastore. It may be assigned a rate limiter.

1. A user creates a Watcher, then enqueues Operations into the Batcher.

1. In the Batcher processing loop, the CapacityInterval asks for capacity from the rate limiter to process the Operations.

1. In the Batcher processing loop, the FlushInterval organizes Operations into batches and raises them back to the Watchers.

1. The user performs the queries as batches are raised.

- There is no logic necessary for NewProvisionResource beyond what is there and nothing to test. There won't be any code coverage for that file and that is the correct design.

- Your right, I will remove that count. It must have been left over from something else.